functional tests with jest-puppeteer (#701)
mdn/yari: testing/README.md

- At the moment it might not work on Windows. It should. At least the "Local
  development"
- We should put all known `kumascript` macros in some form. At least the ones
  intend to support.
- There is no guidelines for how to add tests but feel free to pile on
  sample pages. You should not be afraid to add more.

## Writing headless tests

Headless tests are when we use [`puppeteer`](https://pptr.dev/) to view pages
rendered from the functional tests. If it helps, the non-headless tests
use `fs.readFileSync()` and `cheerio` etc. to inspect the created files in
`client/build/**`.

We use [`jest-puppeteer`](https://github.com/smooth-code/jest-puppeteer) and

> its README is very relevant to help you write tests. And this links to

Reviewer: ```suggestion
its README is very relevant to help you write tests. Here's the link to
```

the [document for `expect-puppeteer`](https://github.com/smooth-code/jest-puppeteer/blob/master/packages/expect-puppeteer/README.md#api)

> which is your best friend writing headless tests.

Reviewer: ```suggestion
which is your best friend when writing headless tests.
```

To get started, open `testing/tests/headless.test.js` and make changes there.
If you need a new page to open, you need to add that to
`testing/content/files/...` first. Then it becomes possible to open it
based on the slug you typed.

Before running the tests, start the function dev server instance:

```sh
yarn start:functional
```

Before you proceed, appreciate that you can now open `http://localhost:5000`
and from there open any page (for example using the search) and what you
see in your browser is what you can expect to see in `jest-puppeteer` in
the tests.

In a separate terminal, run all the tests:

```sh
./testing/scripts/functional-test.sh
```

As you notice, that shell script actually does a lot. It prebuilds the
assets, builds the actual documents, and it runs _all_ `jest` tests.

> To just run, all `jest` tests, just run the last command:

Reviewer: ```suggestion
To just run all `jest` tests, just run the last command:
```

```sh
yarn workspace testing run test
```

Which is just an alias to start `jest` which means you can apply your own
parameters. For example, this starts the `jest` watcher:

```sh
yarn workspace testing run test --watch
```

Once the `jest` watcher has started press <kbd>p</kbd> and type `headless`
and now it only (re-runs) tests the headless tests.

**Note!** that only in local development do you need to start the functional
server first. In GitHub Actions (CI), `jest-puppeteer` is instructed to start
the server as a setup (and teardown) step.

## Debugging headless tests

> What's very likely is that you might be sure what you and how to test

Reviewer: ```suggestion
It's very likely that you'll want to see and test
```
